app/cdap/components/ConnectionList/index.tsx

```diff
@@ -0,0 +1,268 @@
+    }
+  };
+  let connectionId;
+  const [dataForTabs, setDataForTabs] = useState([
+    {
+      data: [],
+      showTabs: true,
+      selectedTab: null,
+      isSearching: false,
+    },
+  ]);
+
+  const getConnectionsTabData = async () => {
+    let connectorTypes = await fetchConnectors();
+    let allConnectionsTotalLength = 0;
+
+    const categorizedConnections = await getCategorizedConnections();
+    connectorTypes = connectorTypes.filter((conn): any => {
+      return [conn.name];
+    });
+
+    connectorTypes = connectorTypes.map((connectorType): any => {
+      const connections = categorizedConnections.get(connectorType.name) || [];
+      allConnectionsTotalLength = allConnectionsTotalLength + connections.length;
+
+      return {
+        ...connectorType,
+        count: connections.length,
+        SVG: <GCSIcon />,
+      };
+    });
+    const firstLevelData = connectorTypes.filter((each) => {
+      if (each.count > 0) {
+        return {
+          name: each.name,
+          count: each.count,
+        };
+      }
+    });
+    setLoading(false);
+    setDataForTabs((prev): any => {
+      const tempData = [...prev];
+      tempData[0].data = firstLevelData;
+      return tempData;
+    });
+  };
+
+  const getCategorizedConnectionsforSelectedTab = async (selectedValue: string, index: number) => {
+    const categorizedConnections = await getCategorizedConnections();
+    const connections = categorizedConnections.get(selectedValue) || [];
+
+    setDataForTabs((prev): any => {
+      const tempData = [...prev];
+      tempData.push({
+        data: [],
+        showTabs: false,
+        selectedTab: '',
+        isSearching: false,
+      });
+      tempData[index + 1][`data`] = connections;
+      tempData[index + 1][`showTabs`] = true;
+      tempData[index + 1][`selectedTab`] = null;
+      return tempData.slice(0, index + 2);
+    });
+    toggleLoader(false);
+  };
+
+  const fetchEntities = async (connectionId, url = pathFromUrl) => {
+    const pathDesired = url ? url : pathFromUrl;
+    const entitiesPromise = exploreConnection({
+      connectionid: connectionId,
+      path: pathDesired,
+    });
+    return entitiesPromise.then((values) => {
+      return values;
+    });
+  };
+
+  const selectedTabValueHandler = (entity: any, index: number) => {
+    toggleLoader(true);
+    setDataForTabs((currentData): any => {
+      let newData = [...currentData];
+      newData[index].selectedTab = entity.name;
+      newData = newData.map((each) => {
+        return {
+          data: each.data,
+          showTabs: each.showTabs,
+          selectedTab: each.selectedTab,
+          isSearching: false,
+        };
+      });
+      if (index === 0) {
+        getCategorizedConnectionsforSelectedTab(entity.name, index);
+      } else if (index === 1) {
+        fetchEntities(entity.name).then((res) => {
+          setDataForTabs((prev): any => {
+            const tempData = [...prev];
```

Comment:
can you make a helper function - it looks like you're doing this multiple times? 

also more comments please theres a lot going on in these functions

Reply:
Done